pandas/core/common.py

```diff
@@ -118,31 +118,24 @@ def is_bool_indexer(key: Any) -> bool:
     check_array_indexer : Check that `key` is a valid array to index,
         and convert to an ndarray.
     """
-    na_msg = "cannot mask with array containing NA / NaN values"
     if isinstance(key, (ABCSeries, np.ndarray, ABCIndex)) or (
         is_array_like(key) and is_extension_array_dtype(key.dtype)
     ):
         if key.dtype == np.object_:
             key = np.asarray(values_from_object(key))
 
             if not lib.is_bool_array(key):
+                na_msg = "Cannot mask with non-boolean array containing NA / NaN values"
                 if isna(key).any():
                     raise ValueError(na_msg)
                 return False
             return True
         elif is_bool_dtype(key.dtype):
-            # an ndarray with bool-dtype by definition has no missing values.
-            # So we only need to check for NAs in ExtensionArrays
-            if is_extension_array_dtype(key.dtype):
-                if np.any(key.isna()):
-                    raise ValueError(na_msg)
             return True
     elif isinstance(key, list):
-        try:
-            arr = np.asarray(key)
-            return arr.dtype == np.bool_ and len(arr) == len(key)
-        except TypeError:  # pragma: no cover
-            return False
+        return all(k is libmissing.NA or k is True or k is False for k in key) and (
```

Comment:
Hmm, this might be much less performant for bigger object dtype arrays with only True/False values? 
Also numpy booleans won't pass the identity check I think.

Reply:
You're right, I'll revert this change

Reply:
Can you think of an alternative approach that might be more performant? The difficulty seems to be treating _only_ `pd.NA` (not `None` or `NaN`) as a valid missing value, so we seemingly can't just rely on `pd.array(key)` having boolean type, for instance